pkg/controller/machine/controller.go

```diff
@@ -177,43 +177,60 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
 	originalConditions := conditions.DeepCopyConditions(m.Status.Conditions)
 
 	if r.gate.Enabled(featuregate.Feature(openshiftfeatures.FeatureGateMachineAPIMigration)) {
-		// Check Status.AuthoritativeAPI
-		// If not MachineAPI. Set the paused condition true and return early.
-		//
-		// Once we have a webhook, we want to remove the check that the AuthoritativeAPI
-		// field is populated.
-		if m.Status.AuthoritativeAPI != "" &&
-			m.Status.AuthoritativeAPI != machinev1.MachineAuthorityMachineAPI {
-			conditions.Set(m, conditions.TrueConditionWithReason(
-				PausedCondition,
-				PausedConditionReason,
-				"The AuthoritativeAPI is set to %s", string(m.Status.AuthoritativeAPI),
-			))
-			if patchErr := r.updateStatus(ctx, m, ptr.Deref(m.Status.Phase, ""), nil, originalConditions); patchErr != nil {
-				klog.Errorf("%v: error patching status: %v", machineName, patchErr)
+		switch m.Status.AuthoritativeAPI {
+		case "":
+			// An empty .status.authoritativeAPI normally means the resource has not yet been reconciled.
+			// and that the value in .spec.authoritativeAPI has not been propagated to .status.authoritativeAPI yet.
+			// This value can be set by two separate controllers, depending which one of them is running at that time,
+			// or in case they are both running, which one gets to set it first (the operation is idempotent so there is no harm in racing).
+			// - the cluster-capi-operator machine-api-migration's migration controller
+			// - this controller
+
+			klog.Infof("%v: machine .status.authoritativeAPI is not yet set, setting it to .spec.authoritativeAPI", m.Name)
+
+			if err := r.patchStatusAuthoritativeAPI(ctx, m, m.Spec.AuthoritativeAPI); err != nil {
+				klog.Errorf("%v: error patching status to set .status.authoritativeAPI for machine: %v", m.Name, err)
+				return reconcile.Result{}, fmt.Errorf("error patching status to set .status.authoritativeAPI for machine %s: %w", m.Name, err)
+			}
```

Comment:
I still worry slightly about duplication of this logic between 2 controllers, especially in separate repos.

* What would happen if we decided to change the 'empty status -> default from spec' logic in the migration controller?
* Can we think of any reason we would ever do that?

I think this is probably ok, just putting on my pessimistic hat to think of problems it could cause.

An alternative to this would be to always allow the migration controller to run, even when the platform is not supported. We could then implement the behaviour for unsupported platforms in one place, which could include defaulting status.authAPI. That feels cleaner to me and it's probably not too much effort, but more than this. Would it have any other benefits to make it worthwhile?

Reply:
[Discussed in slack.](https://redhat-internal.slack.com/archives/GE2HQ9QP4/p1751536658080839)

I'm happy that we've considered the points here.